File: Cpp/Modules/20.Reports/60.Templates On The Exe/UMainForm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "UMainForm.h"
#include <IOUtils.hpp>

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TMainForm *MainForm;
//---------------------------------------------------------------------------
__fastcall TMainForm::TMainForm(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::btnCancelClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------

void __fastcall TMainForm::btnGoClick(TObject *Sender)
{
  RunReport();
}
//---------------------------------------------------------------------------

void __fastcall TMainForm::GetIncludes(TObject* sender, TGetIncludeEventArgs* e)
{
  TResourceStream* IncReport = new TResourceStream((int)HInstance, TPath::GetFileNameWithoutExtension(e->FileName), RT_RCDATA);
  __try
  {
	TBytes b;
	b.Length = IncReport->Size;
	if (b.Length > 0) IncReport->Read(b, b.Length);
	e->IncludeData = b;
  }
  __finally
  {
	IncReport->Free();
  }
}
//---------------------------------------------------------------------------

void __fastcall TMainForm::RunReport()
{
  if (!SaveDialog->Execute()) return;

  TFlexCelReport* Report = new TFlexCelReport(true);
  __try
  {
	Report->GetInclude = GetIncludes; //this is only needed if you have includes.
	Report->AddTable(DemoTables);
	Report->SetValue("Date", TReportValue::Create(TCellValue::Create(Now())));

	TResourceStream* TemplateStream = new TResourceStream((int)HInstance, "TemplatesOnTheExe", RT_RCDATA);
	__try
	{
	  TStream* OutputStream = new TFileStream(SaveDialog->FileName, fmCreate);
	  __try
	  {
		Report->Run(TemplateStream, OutputStream);
	  }
	  __finally
	  {
		OutputStream->Free();
	  }
	}
	__finally
	{
	  TemplateStream->Free();
	}
  }
  __finally
  {
	Report->Free();
  }

 if (MessageDlg(L"Do you want to open the generated file?", mtConfirmation,
	  TMsgDlgButtons() << mbYes << mbNo, 0) == mrYes)
	  {
			ShellExecute(0, L"open", SaveDialog->FileName.c_str(), NULL, NULL,
			SW_SHOWNORMAL);
	  }


}
//---------------------------------------------------------------------------

File: Delphi/Modules/20.Reports/60.Templates In The Exe/UMainForm.pas
unit UMainForm;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics,
  FlexCel.VCLSupport, FlexCel.Core, FlexCel.XlsAdapter, FlexCel.Report, FlexCel.Render,
  {$if CompilerVersion >= 23.0} System.UITypes, {$IFEND}
  ShellApi,
  Controls, Forms, Dialogs, StdCtrls, ExtCtrls;

type
  TMainForm = class(TForm)
    btnCancel: TButton;
    btnGo: TButton;
    SaveDialog: TSaveDialog;
    Label1: TLabel;
    procedure btnCancelClick(Sender: TObject);
    procedure btnGoClick(Sender: TObject);
  private
    procedure RunReport;
    procedure GetIncludes(const sender: TObject;const e: TGetIncludeEventArgs);
    { Private declarations }
  public
    { Public declarations }
  end;

var
  MainForm: TMainForm;

implementation
uses IOUtils, DemoData;

{$R *.dfm}

procedure TMainForm.btnCancelClick(Sender: TObject);
begin
  Close;
end;

procedure TMainForm.btnGoClick(Sender: TObject);
begin
  RunReport;

end;


procedure TMainForm.GetIncludes(const sender: TObject; const e: TGetIncludeEventArgs);
var
  IncReport: TResourceStream;
  b: TBytes;
begin
  IncReport := TResourceStream.Create(hinstance, TPath.GetFileNameWithoutExtension(e.FileName), RT_RCDATA);
  try
    SetLength(b, IncReport.Size);
    if (Length(b) > 0) then IncReport.ReadBuffer(b[0], Length(b));
    e.IncludeData := b;
  finally
    IncReport.Free;
  end;
end;

procedure TMainForm.RunReport;
var
  Report: TFlexCelReport;
  TemplateStream: TResourceStream;
  OutputStream: TFileStream;
begin
  if not SaveDialog.Execute then exit;

  Report := TFlexCelReport.Create(true);
  try
    Report.GetInclude := GetIncludes; //this is only needed if you have includes.
    Report.AddTable(DemoTables);
    Report.SetValue('Date', Now);

    TemplateStream := TResourceStream.Create(hinstance, 'TemplatesInTheExe', RT_RCDATA);
    try
      OutputStream := TFileStream.Create(SaveDialog.FileName, fmCreate);
      try
        Report.Run(TemplateStream, OutputStream);
      finally
        OutputStream.Free;
      end;
    finally
      TemplateStream.Free;
    end;
  finally
    Report.Free;
  end;

  if MessageDlg('Do you want to open the generated file?', mtConfirmation, [mbYes, mbNo], 0) = mrYes then
  begin
    ShellExecute(0, 'open', PCHAR(SaveDialog.FileName), nil, nil, SW_SHOWNORMAL);
  end;


end;

end.
